tls/s2n_client_extensions.c

```diff
@@ -318,8 +319,7 @@ static int s2n_recv_client_alpn(struct s2n_connection *conn, struct s2n_stuffer
 
         GUARD(s2n_stuffer_reread(&client_protos));
     }
-
-    S2N_ERROR(S2N_ERR_NO_APPLICATION_PROTOCOL);
+    return 0;
```

Comment:
From https://tools.ietf.org/html/rfc7301#section-3.1 there is guidance that says we should abort here: 

```
In the event that the server supports no
protocols that the client advertises, then the server SHALL respond
with a fatal "no_application_protocol" alert.
```

However, I think it's better to try and continue and let the application decide what it wants to do. Effectively treat a mismatch of ALPN the same as if ALPN was not offered at all. I'm curious what others think.

Reply:
I agree and think it would be okay to succeed in the event of an ALPN mismatch and do as you suggest by treating it as if ALPN was not offered at all. In that case, I think the server shouldn't respond with the ALPN extension in the ServerHello. The RFC you linked to also says:

>The server will ignore any protocol name that it does not recognize.  A new ServerHello extension type ("application_layer_protocol_negotiation(16)") MAY be returned to the client within the extended ServerHello message.

And if I'm interpreting this correctly, I think that's how this code will work. If conn->application_protocol isn't set, then the extension in the ServerHello won't be sent.